rules-tests/CodeQuality/Rector/Class_/AddReturnDocBlockToCollectionPropertyGetterByToManyAnnotationRector/Fixture/OneToMany/with_index_by.php.inc

```diff
@@ -25,38 +25,3 @@ class Trainer
         $this->trainings = $trainings;
     }
 }
-
-?>
------
-<?php
-
-namespace Rector\Doctrine\Tests\CodeQuality\Rector\Class_\AddReturnDocBlockToCollectionPropertyGetterByToManyAnnotationRector\Fixture\OneToMany;
-
-use Doctrine\Common\Collections\Collection;
-use Doctrine\ORM\Mapping as ORM;
-use Rector\Doctrine\Tests\CodeQuality\Rector\Class_\AddReturnDocBlockToCollectionPropertyGetterByToManyAnnotationRector\Source\Training;
-
-#[ORM\Entity()]
-class Trainer
-{
-    /**
-     * @var Collection|Training[]
-     */
-    #[ORM\OneToMany(targetEntity: Training::class, indexBy: "id", mappedBy: "trainer")]
-    private $trainings = [];
-
-    /**
-     * @return \Doctrine\Common\Collections\Collection<\Rector\Doctrine\Tests\CodeQuality\Rector\Class_\AddReturnDocBlockToCollectionPropertyGetterByToManyAnnotationRector\Source\Training>
```

Comment:
The key should be filled with `string`:

```diff
-@return \Doctrine\Common\Collections\Collection<\Rector\Doctrine\Tests\CodeQuality\Rector\Class_\AddReturnDocBlockToCollectionPropertyGetterByToManyAnnotationRector\Source\Training>
+@return \Doctrine\Common\Collections\Collection<string, \Rector\Doctrine\Tests\CodeQuality\Rector\Class_\AddReturnDocBlockToCollectionPropertyGetterByToManyAnnotationRector\Source\Training>
```

Reply:
The issue is, the lookup require more works, it probably can, but it require lookup target key of id, then lookup target class, then verify its type of property in target class, which may still not property typed, that's why it skipped.

Reply:
For context, the type lookup is looking on target class: `Training` which can be `int`, `string`, or `nullable` when defined nullable: true.

https://github.com/rectorphp/rector-doctrine/blob/1523ce5841f90a9b76c454410d7f6495640f5248/rules-tests/CodeQuality/Rector/Class_/AddReturnDocBlockToCollectionPropertyGetterByToManyAnnotationRector/Source/Training.php#L11-L13

which class Training is gotten from `targetEntity` annotation/attribute definition.

```php
    #[ORM\OneToMany(targetEntity: Training::class, indexBy: "id", mappedBy: "trainer")]
    private $trainings = [];
```

that's why it skipped for this PR.

or just per @AlexeyKosov , the existing use `Collection<ClassName>` when `indexBy` exists is fine as is.

Reply:
@TomasVotruba to get the correct type when `indexBy` exists, it requires steps:

- check Property annotation/attribute `targetEntity` and `indexBy` definition
- on annotation, it may require special value getter for 
      - FQCN::class 
      - string
      - sub string from use 
- utilize `AstResolver` to resolve Target class, if not found, just use integer ?
   - get property with name equal to id definition
   - get annnotation/attribute type of Column
       - another special handling if the type is string, int, or binary?
       - how about nullable?
   - if not found, fallback to integer?

Those complexity is unnecessary imo :)

Reply:
It seems simple. If indexBy has string type property, use `string`. Otherwise fallback to `int`

Reply:
no, `indexBy` is belong to targetEntity, which we need to resolve other class: `Training`, if it defined `id`, we lookup `Training::$id` property, then we read annotation/attribute there.

Reply:
Exactly 💯

Reply:
That back to my comment above, we check annotation/attribute twice https://github.com/rectorphp/rector-doctrine/pull/361#discussion_r1894600569

 with utilize AstResolver, just for it, which imo, is too much

Reply:
Reflection might be enough too. Let's give it a try.

Reply:
Reflection can't read annotation, we still using both annotation and attribute to be read

Reply:
Let's see if I can do it at PR:

- https://github.com/rectorphp/rector-doctrine/pull/362